torchao/modules/nf4_linear.py

```diff
@@ -0,0 +1,35 @@
+import torch
+
+import torch.nn as nn
+from torch import Tensor
+from torchao.dtypes.nf4tensor import NF4Tensor, linear_nf4
+
+
+class FrozenNF4Linear(nn.Linear):
```

Comment:
Just out of curiosity: Why can't this be done by overwriting the weight of an nn.Linear layer?

As in like here

https://github.com/pytorch-labs/segment-anything-fast/blob/387488bc4c7ab2ae311fb0632b34cab5cbfbab78/segment_anything_fast/sparse.py#L28-L32

```
def apply_sparse(model):
    apply_fake_sparsity(model)
    for name, mod in model.named_modules():
        if isinstance(mod, torch.nn.Linear):
            mod.weight = torch.nn.Parameter(to_sparse_semi_structured(mod.weight))
```

Reply:
I think the anwser is cause we don't define nn.linear, see comment above, but if it works then I agree we should

Reply:
For TorchTune use cases, we may not want to overwrite every single linear layer in the model with a frozen NF4 linear. Basically we wanna offer maximum flexibility to users, where even if in QLoRA currently every base linear is overwritten, they might wanna play around with this (for example making more granular tradeoff by only quantizing the qkv projections and not feed forwards). 

Also, this won't work easily because our LoRA adapters are nn.Linears themselves - and we would not want to overwrite the LoRA adapeters. 

cc @ebsmothers for thoughts on UX as well

Reply:
Can you link your existing LoraLinears? I imagine the best UX for users/torchtune would what christian says, and you if you say "I want to qloraify the q projections" you would have a util that swaps the LoraLinear non adapter weight for an nf4 tensor and that should be all that is needed

Reply:
@drisspg Thanks for the suggestion! Here is the `LoRALinear` in torchtune: https://github.com/pytorch-labs/torchtune/blob/2fba15d18d35383f7b8ad4dac5369ca6646ae68e/torchtune/modules/peft/lora.py#L37

I'd imagine such a UX would be -


```
def swap_nf4():
  for module in llama.modules():
      if isinstance(module, LoRALinear):
          # quantize to NF4
          module.weight = NF4.from_tensor(module.weight)
```

Although, as opposed to module / parameter swapping, torchtune prefers to use a componentized builder approach where we build up models such as llama by plugging in the right nn.Module components depending on the config - nn.Linear for regular llama, LoRALinear for LoRA, and now NF4Linear for QLoRA. See an example of the builder pattern here: https://github.com/pytorch-labs/torchtune/blob/main/torchtune/models/llama2/_lora_llama2_builders.py#L135

Reply:
Another point is that IMO `NF4Linear` eliminates a lot of complexity around state_dict save/load. When `load_state_dict`, I'm not sure whether there will be issues loading into a class that uses NF4 tensors. But if we have a specific NF4Linear that uses these NF4Tensros, we can attach load pre and post hooks to upcast / downcast the tensors appropriately.

Reply:
For example, here's what a `load_state_dict` for QLoRA might look like:

```
def load_checkpoint_nf4():
        # NOTE: this also enforces that ALL linear layers will always be quantized with QLoRA.
        # That might not always be the case if users want to customize and for example only
        # quantize some layers.
        # Convert all NF4s to their original weight
        for module in model.modules():
            if isinstance(module, nn.Linear):
                module.weight = module.weight.get_original_weight()
                # would have to add support for bias as well

        load_checkpoint(model)
        # Now re-quantize
        for module in model.modules():
            if isinstance(module, nn.Linear):
                module.weight = NF4Tensor.from_tensor(module.weight.data).to(module.weight.device)
```

This is of course assuming we quantize before loading state_dict, if we quantize after that could bring down the complexity.

Reply:
At least for my mental model it does make sense to inherit from `nn.Linear` here (as opposed to swapping out `self.weight` from an `nn.Linear`, though that is nice and simple). But `FrozenNF4Linear` is basically a constrained version of `nn.Linear`, right? The weight is a particular tensor subclass, and we also require that there be no gradient. So imo we should inherit as a way of being explicit about these constraints. That way I can look at an `FrozenNF4Linear` and know that these conditions should hold, as opposed to having to try and figure them out across all my `nn.Linear`s